Mathlib/Algebra/Group/Action/Equidecomp.lean

```diff
@@ -0,0 +1,215 @@
+(up to the choice of group elements) definition: an *Equidecomposition* of `A` and `B` is a
+bijection `f : A ≃ B` such that for some `S : Finset G`, `f a ∈ S • a` for all `a`.
+
+We take this as our definition as it is easier to work with. It is implemented as an element
+`PartialEquiv X X` with source `A` and target `B`.
+
+## Implementation Notes
+
+* Equidecompositions are implemented as elements of `PartialEquiv X X` together with a
+  `Finset` of elements of the acting group and a proof that every point in the source is moved
+  by an element in the finset.
+
+* The requirement that `G` be a group is relaxed where possible.
+
+* We introduce a non-standard predicate, `IsDecompOn`, to state that a function satisfies the main
+  combinatorial property of equidecompositions, even if it is not injective or surjective.
+
+## TODO
+
+* Prove that if two sets embeddidecompose into eachother, they are equidecomposable
+  (Schroeder-Bernstein type theorem)
+* Define embeddidecomposability as a preorder on sets and prove that its induced equivalence
+  relation is equidecomposability.
+* Prove the definition of equidecomposition used here is equivalent to the more familiar one
+  using partitions.
+
+-/
+
+variable {X G : Type*} {A B C: Set X}
+
+open Function Set Pointwise PartialEquiv
+
+namespace Equidecomp
+
+section SMul
+
+variable [SMul G X]
+
+/-- Let `G` act on a space `X` and `A : Set X`. We say `f : X → X` is a decomposition on `A`
+as witnessed by some `S : Finset G` if for all `a ∈ A`, the value `f a` can be obtained
+by applying some element of `S` to `a` instead.
+
+More familiarly, the restriction of `f` to `A` is the result of partitioning `A` into finitely many
+pieces, then applying a single element of `G` to each piece. -/
+def IsDecompOn (f : X → X) (A : Set X) (S : Finset G) : Prop := ∀ a ∈ A, ∃ g ∈ S, f a = g • a
+
+variable (X G)
+
+/-- Let `G` act on a space `X`. An `Equidecomposition` with respect to `X` and `G` is a partial
+bijection `f : PartialEquiv X X` with the property that for some set `elements : Finset G`,
+(which we record), for each `a ∈ f.source`, `f a` can be obtained by applying some `g ∈ elements`
+instead. We call `f` an equidecomposition of `f.source` with `f.target`.
+
+More familiarly, `f` is the result of partitioning `f.source` into finitely many pieces,
+then applying a single element of `G` to each to get a partition of `f.target`.
+-/
+structure _root_.Equidecomp extends PartialEquiv X X where
+  decomp' : ∃ S : Finset G, IsDecompOn toFun source S
+
+variable {X G}
+
+/-- Note that `Equidecomp X G` is not `FunLike`. -/
+instance : CoeFun (Equidecomp X G) fun _ => X → X := ⟨fun f => f.toFun⟩
+
+/-- A finite set of group elements witnessing that `f` is an equidecomposition. -/
+noncomputable
+def witness (f : Equidecomp X G) : Finset G := f.decomp'.choose
+
+theorem decomp (f : Equidecomp X G) : IsDecompOn f f.source f.witness := f.decomp'.choose_spec
```

Comment:
```suggestion
theorem isDecompOn_source_withness (f : Equidecomp X G) : IsDecompOn f f.source f.witness :=
  f.decomp'.choose_spec
```

Reply:
I think it is worth keeping this name short since it is the key property of an equidecomposition and will be invoked repeatedly when working with them. I like `Equidecomp.isDecompOn`